services/s3/src/main/java/software/amazon/awssdk/services/s3/internal/endpoints/S3EndpointResolverFactory.java

```diff
@@ -0,0 +1,39 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.services.s3.internal.endpoints;
+
+import java.util.Optional;
+import software.amazon.awssdk.annotations.SdkInternalApi;
+
+/**
+ * Get endpoint resolver.
+ */
+@SdkInternalApi
+public final class S3EndpointResolverFactory {
+
+    private static final S3EndpointResolver ACCESS_POINT_ENDPOINT_RESOLVER = S3AccessPointEndpointResolver.create();
+    private static final S3EndpointResolver BUCKET_ENDPOINT_RESOLVER = S3BucketEndpointResolver.create();
+
+    private S3EndpointResolverFactory() {
+    }
+
+    public static S3EndpointResolver getEndpointResolver(String bucketName) {
+        return Optional.ofNullable(bucketName)
+                       .filter(S3EndpointUtils::isArn)
+                       .map(arn -> ACCESS_POINT_ENDPOINT_RESOLVER)
+                       .orElseGet(() -> BUCKET_ENDPOINT_RESOLVER);
```

Comment:
I'm wondering if it would be more readable just to say
```
if (bucketName != null && isArn(bucketName) { 
    return ACCESS_POINT_ENDPOINT_RESOLVER; 
} 
return BUCKET_ENDPOINT RESOLVER;
```

Reply:
Oh come on! Stomp on my nice functional code, will you?